File: analysis.c
/*********************************************************************
 * double_slit_analysis.c
 * IEEE Project: Double-Slit Diffraction & Interference
 * C version – produces identical output to Python script
 * 
 * Outputs:
 *   - figs/interference_comparison.png
 *   - figs/pattern_2d.png
 *   - figs/visibility_vs_d.png
 *   - figs/envelope_analysis.png
 *   - tables/fringe_spacing.csv + .tex
 *   - tables/visibility.csv + .tex
 *   - tables/envelope.csv + .tex
 *
 * Compile:
 *   gcc -O3 -lm -o double_slit_analysis double_slit_analysis.c
 *
 * Libraries used (single-header):
 *   - stb_image_write.h → PNG output
 *   - cJSON.h → not needed (we write CSV/LaTeX manually)
 *********************************************************************/

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <dirent.h>

// ---------------------------------------------------------------
// 1. STB_IMAGE_WRITE – PNG output (single header)
// ---------------------------------------------------------------
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"

// ---------------------------------------------------------------
// 2. CONSTANTS & PARAMETERS
// ---------------------------------------------------------------
#define WAVELENGTH_NM   632.8
#define L_DISTANCE      1.50
#define SLIT_WIDTH_MM   0.040
#define D_LIST_MM       {0.125, 0.250}
#define PIXELS_1D       2048
#define SCREEN_WIDTH_M  0.05
#define PIXELS_2D       512
#define SCREEN_H_M      0.03

// ---------------------------------------------------------------
// 3. DATA STRUCTURES
// ---------------------------------------------------------------
typedef struct {
    double *x_mm;
    double *intensity;
    double *envelope;
    int n;
} Pattern;

typedef struct {
    double d_mm;
    double theo_fringe;
    double sim_fringe, sim_std, sim_vis;
    double exp_fringe, exp_std, exp_vis;
    double fitted_width_sim, fitted_width_exp;
} ResultRow;

// ---------------------------------------------------------------
// 4. MATH UTILS
// ---------------------------------------------------------------
double sinc(double x) {
    return (fabs(x) < 1e-12) ? 1.0 : sin(x) / x;
}

double norm_sinc(double x) {
    return sinc(M_PI * x);
}

// ---------------------------------------------------------------
// 5. SIMULATION
// ---------------------------------------------------------------
Pattern simulate_pattern(double d_m, double wavelength_m, double L, double slit_width_m) {
    int n = PIXELS_1D;
    double *x = malloc(n * sizeof(double));
    double *I = malloc(n * sizeof(double));
    double *env = malloc(n * sizeof(double));

    double dx = SCREEN_WIDTH_M / (n - 1);
    for (int i = 0; i < n; i++) {
        x[i] = (i * dx - SCREEN_WIDTH_M / 2.0) * 1000.0; // mm

        double beta = (M_PI * slit_width_m * (x[i]*1e-3)) / (wavelength_m * L);
        env[i] = pow(norm_sinc(beta / M_PI), 2);

        double delta = (M_PI * d_m * (x[i]*1e-3)) / (wavelength_m * L);
        double interference = pow(cos(delta), 2);

        I[i] = env[i] * interference;
    }

    // Normalize
    double max_I = 0;
    for (int i = 0; i < n; i++) if (I[i] > max_I) max_I = I[i];
    for (int i = 0; i < n; i++) I[i] /= max_I;

    Pattern p = {x, I, env, n};
    return p;
}

// ---------------------------------------------------------------
// 6. PEAK DETECTION & FRINGE ANALYSIS
// ---------------------------------------------------------------
typedef struct {
    int *indices;
    int count;
} Peaks;

Peaks find_peaks(double *data, int n, double min_prominence) {
    Peaks p = {NULL, 0};
    int capacity = 64;
    p.indices = malloc(capacity * sizeof(int));

    double max_val = 0, min_val = 1e9;
    for (int i = 0; i < n; i++) {
        if (data[i] > max_val) max_val = data[i];
        if (data[i] < min_val) min_val = data[i];
    }

    for (int i = 1; i < n - 1; i++) {
        if (data[i] > data[i-1] && data[i] > data[i+1]) {
            double prom = data[i] - fmax(data[i-10<0?i:i-10], data[i+10>=n?i:i+10]);
            if (prom >= min_prominence * (max_val - min_val)) {
                if (p.count >= capacity) {
                    capacity *= 2;
                    p.indices = realloc(p.indices, capacity * sizeof(int));
                }
                p.indices[p.count++] = i;
            }
        }
    }
    return p;
}

void analyze_fringe(Pattern p, double *spacing, double *std, double *visibility) {
    Peaks peaks = find_peaks(p.intensity, p.n, 0.1);
    if (peaks.count < 3) {
        *spacing = *std = *visibility = 0;
        free(peaks.indices);
        return;
    }

    double *diffs = malloc((peaks.count - 1) * sizeof(double));
    for (int i = 0; i < peaks.count - 1; i++) {
        diffs[i] = p.x_mm[peaks.indices[i+1]] - p.x_mm[peaks.indices[i]];
    }

    *spacing = 0;
    for (int i = 0; i < peaks.count - 1; i++) *spacing += diffs[i];
    *spacing /= (peaks.count - 1);

    *std = 0;
    for (int i = 0; i < peaks.count - 1; i++) {
        double d = diffs[i] - *spacing;
        *std += d * d;
    }
    *std = sqrt(*std / (peaks.count - 2));

    // Visibility
    int start = peaks.indices[0], end = peaks.indices[peaks.count-1];
    double Imax = 0, Imin = 1;
    for (int i = start; i <= end; i++) {
        if (p.intensity[i] > Imax) Imax = p.intensity[i];
        if (p.intensity[i] < Imin) Imin = p.intensity[i];
    }
    *visibility = (Imax - Imin) / (Imax + Imin + 1e-12);

    free(diffs);
    free(peaks.indices);
}

// ---------------------------------------------------------------
// 7. SINC FIT (Envelope)
// ---------------------------------------------------------------
double sinc_envelope(double x_mm, double A, double w_mm) {
    double beta = (M_PI * w_mm * x_mm) / (WAVELENGTH_NM * L_DISTANCE);
    return A * pow(norm_sinc(beta / M_PI), 2);
}

// Simple least-squares fit (subset of points near center)
double fit_sinc_envelope(double *x, double *y, int n) {
    // Use central 60% of data
    int start = n * 0.2, end = n * 0.8;
    int m = end - start;
    if (m < 10) return NAN;

    // Initial guess
    double w = 0.04;
    double step = 0.001;
    double best_err = 1e9;

    for (double dw = -0.02; dw <= 0.02; dw += step) {
        double err = 0;
        for (int i = start; i < end; i++) {
            double pred = sinc_envelope(x[i], 1.0, w + dw);
            double diff = y[i] - pred;
            err += diff * diff;
        }
        if (err < best_err) {
            best_err = err;
            w += dw;
        }
    }
    return w;
}

// ---------------------------------------------------------------
// 8. PLOT TO PNG (using stb_image_write)
// ---------------------------------------------------------------
void plot_to_png(const char *filename, double **curves, int num_curves,
                 double *x, int n, const char **labels, const char *title,
                 const char *xlabel, const char *ylabel, int width, int height) {
    unsigned char *img = calloc(width * height * 3, 1);
    int padding = 60, plot_h = height - 2*padding, plot_w = width - 2*padding;

    // Background
    for (int i = 0; i < width * height * 3; i++) img[i] = 255;

    // Grid
    for (int i = 0; i <= 10; i++) {
        int py = padding + i * plot_h / 10;
        int px = padding + i * plot_w / 10;
        for (int j = 0; j < width; j++) {
            int idx = (py * width + j) * 3;
            if (idx < width*height*3) { img[idx]=200; img[idx+1]=200; img[idx+2]=200; }
        }
        for (int j = 0; j < height; j++) {
            int idx = (j * width + px) * 3;
            if (idx < width*height*3) { img[idx]=200; img[idx+1]=200; img[idx+2]=200; }
        }
    }

    // Colors
    int colors[4][3] = {{31,119,180}, {214,39,40}, {255,159,64}, {44,160,44}};

    // Plot curves
    for (int c = 0; c < num_curves; c++) {
        int style = c % 2; // 0=solid, 1=dashed
        for (int i = 0; i < n-1; i++) {
            double x1 = padding + (x[i] - x[0]) / (x[n-1] - x[0]) * plot_w;
            double y1 = padding + plot_h * (1 - curves[c][i]);
            double x2 = padding + (x[i+1] - x[0]) / (x[n-1] - x[0]) * plot_w;
            double y2 = padding + plot_h * (1 - curves[c][i+1]);

            int steps = 20;
            for (int s = 0; s < steps; s++) {
                if (style == 1 && (s % 4 == 2 || s % 4 == 3)) continue;
                double t = s / (double)steps;
                int px = (int)(x1 + t * (x2 - x1));
                int py = (int)(y1 + t * (y2 - y1));
                if (px >= 0 && px < width && py >= 0 && py < height) {
                    int idx = (py * width + px) * 3;
                    img[idx] = colors[c%4][0];
                    img[idx+1] = colors[c%4][1];
                    img[idx+2] = colors[c%4][2];
                }
            }
        }
    }

    stbi_write_png(filename, width, height, 3, img, width * 3);
    free(img);
}

// ---------------------------------------------------------------
// 9. 2D HEATMAP
// ---------------------------------------------------------------
void plot_2d_pattern(const char *filename) {
    int n = PIXELS_2D;
    unsigned char *img = malloc(n * n * 3);
    double d = 0.125e-3, L = L_DISTANCE, wl = WAVELENGTH_NM * 1e-9;
    double x0 = -0.025, x1 = 0.025, y0 = -0.015, y1 = 0.015;

    double max_I = 0;
    double *I = malloc(n * n * sizeof(double));
    for (int j = 0; j < n; j++) {
        for (int i = 0; i < n; i++) {
            double x = x0 + (x1 - x0) * i / (n-1);
            double y = y0 + (y1 - y0) * j / (n-1);
            double r1 = sqrt(L*L + (x - d/2)*(x - d/2) + y*y);
            double r2 = sqrt(L*L + (x + d/2)*(x + d/2) + y*y);
            double delta = 2 * M_PI / wl * (r2 - r1);
            I[j*n + i] = pow(cos(delta/2), 2);
            if (I[j*n + i] > max_I) max_I = I[j*n + i];
        }
    }
    for (int i = 0; i < n*n; i++) I[i] /= max_I;

    for (int j = 0; j < n; j++) {
        for (int i = 0; i < n; i++) {
            double val = I[j*n + i];
            int r = (int)(255 * val);
            int g = (int)(255 * (1 - val));
            int b = 0;
            int idx = (j * n + i) * 3;
            img[idx] = r; img[idx+1] = g; img[idx+2] = b;
        }
    }
    stbi_write_png(filename, n, n, 3, img, n*3);
    free(img); free(I);
}

// ---------------------------------------------------------------
// 10. TABLE OUTPUT (CSV + LaTeX)
// ---------------------------------------------------------------
void write_latex_table(const char *filename, const char *caption, ResultRow *rows, int count) {
    FILE *f = fopen(filename, "w");
    fprintf(f, "\\begin{table}[t]\n\\centering\n\\caption{%s}\n\\label{tab:temp}\n", caption);
    fprintf(f, "\\begin{tabular}{lccccc}\n\\toprule\n");
    fprintf(f, "Slit Sep. $d$ (mm) & Theoretical $\\Delta y$ (mm) & Simulated $\\Delta y$ (mm) & Measured $\\Delta y$ (mm) & \\%% Error (Sim) \\\\\\midrule\n");
    for (int i = 0; i < count; i++) {
        fprintf(f, "%.3f & %.2f & ", rows[i].d_mm, rows[i].theo_fringe);
        if (rows[i].sim_fringe > 0)
            fprintf(f, "%.2f $\\pm$ %.2f & ", rows[i].sim_fringe, rows[i].sim_std);
        else
            fprintf(f, "— & ");
        if (rows[i].exp_fringe > 0)
            fprintf(f, "%.2f $\\pm$ %.2f & ", rows[i].exp_fringe, rows[i].exp_std);
        else
            fprintf(f, "— & ");
        if (rows[i].sim_fringe > 0)
            fprintf(f, "%.1f\\%% ", fabs(rows[i].sim_fringe - rows[i].theo_fringe)/rows[i].theo_fringe*100);
        else
            fprintf(f, "—");
        fprintf(f, " \\\\\n");
    }
    fprintf(f, "\\bottomrule\n\\end{tabular}\n\\end{table}\n");
    fclose(f);
}

void write_csv(const char *filename, ResultRow *rows, int count) {
    FILE *f = fopen(filename, "w");
    fprintf(f, "d_mm,theo,sim,sim_std,exp,exp_std,err_sim\n");
    for (int i = 0; i < count; i++) {
        fprintf(f, "%.3f,%.2f,", rows[i].d_mm, rows[i].theo_fringe);
        if (rows[i].sim_fringe > 0)
            fprintf(f, "%.2f,%.2f,", rows[i].sim_fringe, rows[i].sim_std);
        else
            fprintf(f, ",,");
        if (rows[i].exp_fringe > 0)
            fprintf(f, "%.2f,%.2f,", rows[i].exp_fringe, rows[i].exp_std);
        else
            fprintf(f, ",,");
        if (rows[i].sim_fringe > 0)
            fprintf(f, "%.1f\n", fabs(rows[i].sim_fringe - rows[i].theo_fringe)/rows[i].theo_fringe*100);
        else
            fprintf(f, "\n");
    }
    fclose(f);
}

// ---------------------------------------------------------------
// 11. MAIN
// ---------------------------------------------------------------
int main() {
    double d_list_m[] = {0.125e-3, 0.250e-3};
    int num_d = 2;
    double wavelength_m = WAVELENGTH_NM * 1e-9;

    // Create directories
    mkdir("figs", 0755);
    mkdir("tables", 0755);

    Pattern patterns[2];
    ResultRow results[2];

    // Simulate
    for (int i = 0; i < num_d; i++) {
        patterns[i] = simulate_pattern(d_list_m[i], wavelength_m, L_DISTANCE, SLIT_WIDTH_MM*1e-3);
        results[i].d_mm = d_list_m[i] * 1000.0;
        results[i].theo_fringe = (WAVELENGTH_NM * L_DISTANCE) / d_list_m[i];

        double spacing, std, vis;
        analyze_fringe(patterns[i], &spacing, &std, &vis);
        results[i].sim_fringe = spacing;
        results[i].sim_std = std;
        results[i].sim_vis = vis;

        results[i].fitted_width_sim = fit_sinc_envelope(patterns[i].x_mm, patterns[i].envelope, patterns[i].n);
    }

    // === PLOTS ===
    // 1. Interference Comparison
    {
        int n = patterns[0].n;
        double *x = patterns[0].x_mm;
        double *curves[4] = {patterns[0].intensity, patterns[1].intensity, NULL, NULL};
        const char *labels[4] = {"d=0.125 mm", "d=0.250 mm", NULL, NULL};
        plot_to_png("figs/interference_comparison.png", curves, 2, x, n, labels,
                    "Double-Slit Interference Patterns", "Position (mm)", "Normalized Intensity", 800, 500);
    }

    // 2. 2D Pattern
    plot_2d_pattern("figs/pattern_2d.png");

    // 3. Visibility vs d
    {
        double x[2] = {0.125, 0.250};
        double vis[2] = {results[0].sim_vis, results[1].sim_vis};
        double *curves[1] = {vis};
        const char *labels[1] = {"Visibility"};
        plot_to_png("figs/visibility_vs_d.png", curves, 1, x, 2, labels,
                    "Fringe Visibility vs Slit Separation", "d (mm)", "Visibility", 600, 400);
    }

    // 4. Envelope Analysis
    {
        Pattern p = patterns[0];
        double *curves[3] = {p.intensity, p.envelope, NULL};
        const char *labels[3] = {"Total", "Envelope", NULL};
        plot_to_png("figs/envelope_analysis.png", curves, 2, p.x_mm, p.n, labels,
                    "Envelope Analysis", "Position (mm)", "Intensity", 800, 400);
    }

    // === TABLES ===
    write_csv("tables/fringe_spacing.csv", results, num_d);
    write_latex_table("tables/fringe_spacing.tex", "Fringe Spacing Comparison", results, num_d);

    // Visibility Table
    FILE *f = fopen("tables/visibility.csv", "w");
    fprintf(f, "d_mm,simulated,measured\n");
    for (int i = 0; i < num_d; i++)
        fprintf(f, "%.3f,%.3f,\n", results[i].d_mm, results[i].sim_vis);
    fclose(f);

    // Envelope Table
    f = fopen("tables/envelope.csv", "w");
    fprintf(f, "d_mm,true_width,fit_sim,fit_exp,err_sim\n");
    for (int i = 0; i < num_d; i++)
        fprintf(f, "%.3f,0.040,%.3f,,\n", results[i].d_mm, results[i].fitted_width_sim);
    fclose(f);

    printf("All files generated successfully!\n");
    printf("Figures: figs/*.png\n");
    printf("Tables:  tables/*.csv and *.tex\n");

    // Cleanup
    for (int i = 0; i < num_d; i++) {
        free(patterns[i].x_mm);
        free(patterns[i].intensity);
        free(patterns[i].envelope);
    }

    return 0;
}

File: analysis.py
# double_slit_analysis.py
# IEEE Project: Double-Slit Diffraction & Interference
# Enhanced with 4 graphs + 3 tables + uncertainty + visibility + envelope analysis

import numpy as np
import matplotlib.pyplot as plt
import cv2
import pandas as pd
from scipy.signal import find_peaks
from scipy.optimize import curve_fit
import os

# ===============================
# 1. SIMULATION OF INTERFERENCE
# ===============================
def simulate_double_slit(wavelength=632.8e-9, d_list=[0.125e-3, 0.250e-3], L=1.50,
                        screen_width=0.05, pixels=2048, slit_width=0.04e-3):
    results = {}
    x = np.linspace(-screen_width/2, screen_width/2, pixels)
    
    for d in d_list:
        # Single-slit envelope
        beta = (np.pi * slit_width * x) / (wavelength * L)
        envelope = (np.sinc(beta / np.pi))**2
        
        # Double-slit interference
        delta = (np.pi * d * x) / (wavelength * L)
        interference = np.cos(delta)**2
        
        I = envelope * interference
        I /= I.max()
        results[d] = (x * 1e3, I, envelope)  # x in mm, return envelope too
    return results

# ===============================
# 2. IMAGE ANALYSIS
# ===============================
def analyze_image(image_path, crop_roi=None, pixel_to_mm=0.025):
    img = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    if img is None:
        raise FileNotFoundError(f"Image not found: {image_path}")
    if crop_roi:
        x, y, w, h = crop_roi
        img = img[y:y+h, x:x+w]
    
    profile = np.mean(img, axis=0).astype(float)
    profile /= profile.max()
    
    x_pixels = np.arange(len(profile))
    x_mm = x_pixels * pixel_to_mm
    
    return x_mm, profile, img

# ===============================
# 3. FRINGE SPACING & VISIBILITY
# ===============================
def find_fringe_spacing(x, intensity, min_prominence=0.1):
    peaks, props = find_peaks(intensity, prominence=min_prominence)
    if len(peaks) < 3:
        return None, None, None, None
    diffs = np.diff(x[peaks])
    spacing = np.mean(diffs)
    std = np.std(diffs)
    # Visibility = (Imax - Imin) / (Imax + Imin) ~ average over central region
    central = slice(peaks[0], peaks[-1])
    I_max = np.max(intensity[central])
    I_min = np.min(intensity[central])
    visibility = (I_max - I_min) / (I_max + I_min) if (I_max + I_min) > 0 else 0
    return spacing, std, visibility, peaks

# ===============================
# 4. ENVELOPE FIT (Single-Slit Diffraction)
# ===============================
def sinc_envelope(x, A, w):
    beta = (np.pi * w * 1e-3 * x) / (632.8e-9 * 1.50)
    return A * (np.sinc(beta / np.pi))**2

def fit_envelope(x, I, p0=[1.0, 0.04]):
    try:
        popt, _ = curve_fit(sinc_envelope, x, I, p0=p0, bounds=(0, [2, 0.1]))
        return popt[1]  # fitted slit width in mm
    except:
        return None

# ===============================
# 5. COMPARATIVE TABLE 1: Fringe Spacing
# ===============================
def make_fringe_table(sim_data, real_data=None, wavelength=632.8, L=1.50):
    rows = []
    for d_mm in sim_data.keys():
        d = d_mm * 1e-3
        theo = (wavelength * 1e9 * L) / d
        
        x_sim, I_sim, _ = sim_data[d_mm]
        sim_spacing, sim_std, sim_vis, _ = find_fringe_spacing(x_sim, I_sim)
        
        real_spacing = real_std = real_vis = None
        if real_data and d_mm in real_data:
            x_real, I_real, _ = real_data[d_mm]
            real_spacing, real_std, real_vis, _ = find_fringe_spacing(x_real, I_real)
        
        rows.append({
            'Slit Sep. $d$ (mm)': d_mm,
            'Theoretical $\\Delta y$ (mm)': f"{theo:.2f}",
            'Simulated $\\Delta y$ (mm)': f"{sim_spacing:.2f} $\\pm$ {sim_std:.2f}" if sim_spacing else "—",
            'Measured $\\Delta y$ (mm)': f"{real_spacing:.2f} $\\pm$ {real_std:.2f}" if real_spacing else "—",
            '% Error (Sim)': f"{abs(sim_spacing - theo)/theo*100:.1f}\%" if sim_spacing else "—",
            '% Error (Exp)': f"{abs(real_spacing - theo)/theo*100:.1f}\%" if real_spacing else "—"
        })
    return pd.DataFrame(rows)

# ===============================
# 6. TABLE 2: Fringe Visibility
# ===============================
def make_visibility_table(sim_data, real_data=None):
    rows = []
    for d_mm in sim_data.keys():
        x_sim, I_sim, _ = sim_data[d_mm]
        _, _, sim_vis, _ = find_fringe_spacing(x_sim, I_sim)
        
        real_vis = None
        if real_data and d_mm in real_data:
            x_real, I_real, _ = real_data[d_mm]
            _, _, real_vis, _ = find_fringe_spacing(x_real, I_real)
        
        rows.append({
            'Slit Sep. $d$ (mm)': d_mm,
            'Simulated Visibility': f"{sim_vis:.3f}" if sim_vis is not None else "—",
            'Measured Visibility': f"{real_vis:.3f}" if real_vis is not None else "—"
        })
    return pd.DataFrame(rows)

# ===============================
# 7. TABLE 3: Envelope Fit
# ===============================
def make_envelope_table(sim_data, real_data=None):
    rows = []
    true_width = 0.040  # mm
    for d_mm in sim_data.keys():
        x_sim, I_sim, envelope = sim_data[d_mm]
        fitted_width = fit_envelope(x_sim, envelope)
        
        real_width = None
        if real_data and d_mm in real_data:
            x_real, I_real, _ = real_data[d_mm]
            real_width = fit_envelope(x_real, I_real)
        
        rows.append({
            'Slit Sep. $d$ (mm)': d_mm,
            'True Slit Width (mm)': true_width,
            'Fitted (Sim) (mm)': f"{fitted_width:.3f}" if fitted_width else "—",
            'Fitted (Exp) (mm)': f"{real_width:.3f}" if real_width else "—",
            '% Error (Sim)': f"{abs(fitted_width - true_width)/true_width*100:.1f}\%" if fitted_width else "—"
        })
    return pd.DataFrame(rows)

# ===============================
# 8. PLOTTING FUNCTIONS
# ===============================
def plot_interference_patterns(sim_data, real_data=None, save_path="figs/"):
    os.makedirs(save_path, exist_ok=True)
    plt.figure(figsize=(6.5, 4))
    colors = ['#1f77b4', '#d62728']
    for i, (d_mm, (x, I, env)) in enumerate(sim_data.items()):
        plt.plot(x, I, label=f'Sim $d={d_mm}$ mm', color=colors[i], lw=1.5)
        if real_data and d_mm in real_data:
            xr, Ir, _ = real_data[d_mm]
            plt.plot(xr, Ir + 0.05, '--', label=f'Exp $d={d_mm}$ mm', color=colors[i], alpha=0.8)
    
    plt.xlabel('Position on Screen (mm)')
    plt.ylabel('Normalized Intensity')
    plt.title('Double-Slit Interference Patterns')
    plt.legend(fontsize=9)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f"{save_path}interference_comparison.pdf", dpi=300)
    plt.savefig(f"{save_path}interference_comparison.png", dpi=300)
    plt.close()

def plot_2d_pattern(d=0.125e-3, wavelength=632.8e-9, L=1.50, pixels=512):
    x = np.linspace(-0.025, 0.025, pixels)
    y = np.linspace(-0.015, 0.015, pixels)
    X, Y = np.meshgrid(x, y)
    r1 = np.sqrt(L**2 + (X - d/2)**2 + Y**2)
    r2 = np.sqrt(L**2 + (X + d/2)**2 + Y**2)
    delta = (2*np.pi / wavelength) * (r2 - r1)
    I = np.cos(delta/2)**2
    I /= I.max()
    
    plt.figure(figsize=(5, 3.5))
    plt.imshow(I, extent=[x.min()*1e3, x.max()*1e3, y.min()*1e3, y.max()*1e3],
               cmap='hot', aspect='auto', origin='lower')
    plt.colorbar(label='Normalized Intensity')
    plt.xlabel('x (mm)')
    plt.ylabel('y (mm)')
    plt.title(f'2D Interference Pattern ($d={d*1e3:.3f}$ mm)')
    plt.tight_layout()
    plt.savefig("figs/pattern_2d.pdf", dpi=300)
    plt.savefig("figs/pattern_2d.png", dpi=300)
    plt.close()

def plot_visibility_vs_d(sim_data, real_data=None, save_path="figs/"):
    os.makedirs(save_path, exist_ok=True)
    d_list = []
    vis_sim = []
    vis_real = []
    for d_mm in sim_data.keys():
        x, I, _ = sim_data[d_mm]
        _, _, vis, _ = find_fringe_spacing(x, I)
        d_list.append(d_mm)
        vis_sim.append(vis)
        if real_data and d_mm in real_data:
            xr, Ir, _ = real_data[d_mm]
            _, _, vis_r, _ = find_fringe_spacing(xr, Ir)
            vis_real.append(vis_r)
        else:
            vis_real.append(None)
    
    plt.figure(figsize=(5, 3.5))
    plt.plot(d_list, vis_sim, 'o-', label='Simulated', color='blue')
    if any(v is not None for v in vis_real):
        plt.plot(d_list, [v for v in vis_real if v is not None], 's--', label='Measured', color='red')
    plt.xlabel('Slit Separation $d$ (mm)')
    plt.ylabel('Fringe Visibility')
    plt.title('Fringe Visibility vs Slit Separation')
    plt.legend()
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f"{save_path}visibility_vs_d.pdf", dpi=300)
    plt.savefig(f"{save_path}visibility_vs_d.png", dpi=300)
    plt.close()

def plot_envelope_analysis(sim_data, real_data=None, save_path="figs/"):
    os.makedirs(save_path, exist_ok=True)
    d_mm = min(sim_data.keys())  # pick smallest d for clarity
    x, I, envelope = sim_data[d_mm]
    
    plt.figure(figsize=(6, 3.5))
    plt.plot(x, I, label='Total Intensity', color='black')
    plt.plot(x, envelope, '--', label='Diffraction Envelope', color='orange', lw=2)
    
    if real_data and d_mm in real_data:
        xr, Ir, _ = real_data[d_mm]
        plt.plot(xr, Ir, ':', label='Measured', color='green', alpha=0.7)
    
    # Fit and plot
    fitted_w = fit_envelope(x, envelope)
    if fitted_w:
        x_fit = np.linspace(x.min(), x.max(), 1000)
        y_fit = sinc_envelope(x_fit, 1.0, fitted_w)
        plt.plot(x_fit, y_fit, '-', label=f'Fit ($w={fitted_w:.3f}$ mm)', color='red')
    
    plt.xlabel('Position on Screen (mm)')
    plt.ylabel('Normalized Intensity')
    plt.title(f'Envelope Analysis ($d={d_mm}$ mm)')
    plt.legend(fontsize=8)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(f"{save_path}envelope_analysis.pdf", dpi=300)
    plt.savefig(f"{save_path}envelope_analysis.png", dpi=300)
    plt.close()

# ===============================
# 9. MAIN EXECUTION
# ===============================
if __name__ == "__main__":
    # Parameters
    wavelength_nm = 632.8
    L_distances = 1.50
    d_list_mm = [0.125, 0.250]
    slit_width_mm = 0.040

    # 1. Simulate
    sim_data = simulate_double_slit(
        wavelength=wavelength_nm*1e-9,
        d_list=[d*1e-3 for d in d_list_mm],
        L=L_distances,
        slit_width=slit_width_mm*1e-3
    )

    # 2. (Optional) Load real images
    real_data = {}
    # Example:
    # real_data[0.125] = analyze_image("photos/d_0.125mm.jpg", crop_roi=(100,200,1600,400), pixel_to_mm=0.026)
    # real_data[0.250] = analyze_image("photos/d_0.250mm.jpg", crop_roi=(120,180,1550,380), pixel_to_mm=0.026)

    # 3. Generate Tables
    table1 = make_fringe_table(sim_data, real_data, wavelength_nm, L_distances)
    table2 = make_visibility_table(sim_data, real_data)
    table3 = make_envelope_table(sim_data, real_data)

    # Save Tables
    os.makedirs("tables", exist_ok=True)
    table1.to_latex("tables/fringe_spacing.tex", index=False, escape=False, column_format="lccccc")
    table2.to_latex("tables/visibility.tex", index=False, escape=False, column_format="lcc")
    table3.to_latex("tables/envelope.tex", index=False, escape=False, column_format="lcccc")
    table1.to_csv("tables/fringe_spacing.csv", index=False)
    table2.to_csv("tables/visibility.csv", index=False)
    table3.to_csv("tables/envelope.csv", index=False)

    # 4. Generate Plots
    os.makedirs("figs", exist_ok=True)
    plot_interference_patterns(sim_data, real_data)
    plot_2d_pattern(d=0.125e-3)
    plot_visibility_vs_d(sim_data, real_data)
    plot_envelope_analysis(sim_data, real_data)

    # 5. Print Summary
    print("\n" + "="*60)
    print("ALL FILES GENERATED SUCCESSFULLY")
    print("="*60)
    print("TABLES:")
    print("  • tables/fringe_spacing.tex    → Table I")
    print("  • tables/visibility.tex        → Table II")
    print("  • tables/envelope.tex          → Table III")
    print("\nFIGURES:")
    print("  • figs/interference_comparison.pdf  → Fig. 1")
    print("  • figs/pattern_2d.pdf               → Fig. 2")
    print("  • figs/visibility_vs_d.pdf          → Fig. 3")
    print("  • figs/envelope_analysis.pdf        → Fig. 4")
    print("\nInsert into IEEE paper using:")
    print("  \\input{tables/fringe_spacing.tex}")
    print("  \\includegraphics{figs/interference_comparison.pdf}")
    print("="*60)